File: AlgorithmConcepts/1_Two_Pointers/c++/Problem_LeetCode-524.-Longest-Word-in-Dictionary-through-Deleting/solution.cpp
#include <string>
#include <vector>

class Solution {
public:
    // LeetCode 524. Longest Word in Dictionary through Deleting
    std::string findLongestWord(const std::string& s, std::vector<std::string>& dictionary) {
        std::string best;
        for (const std::string& word : dictionary) {
            if (isSubsequence(word, s)) {
                if (word.size() > best.size() || (word.size() == best.size() && word < best)) {
                    best = word;
                }
            }
        }
        return best;
    }

private:
    static bool isSubsequence(const std::string& word, const std::string& source) {
        std::size_t i = 0;
        for (char ch : source) {
            if (i < word.size() && word[i] == ch) {
                ++i;
                if (i == word.size()) {
                    return true;
                }
            }
        }
        return i == word.size();
    }
};

File: AlgorithmConcepts/1_Two_Pointers/python/Problem_LeetCode-524.-Longest-Word-in-Dictionary-through-Deleting/solution.py
from typing import List


class Solution:
    """LeetCode 524. Longest Word in Dictionary through Deleting.

    For each candidate word we perform a linear scan against ``s`` to determine whether
    it forms a subsequence. We track the best match by preferring longer words and
    breaking ties with lexicographical order as required by the problem statement.
    """

    def findLongestWord(self, s: str, dictionary: List[str]) -> str:
        """Return the longest dictionary word that is a subsequence of ``s``."""

        def is_subsequence(word: str) -> bool:
            i = 0
            for char in s:
                if i < len(word) and word[i] == char:
                    i += 1
                    if i == len(word):
                        return True
            return i == len(word)

        best = ""
        for word in dictionary:
            if is_subsequence(word):
                if len(word) > len(best) or (len(word) == len(best) and word < best):
                    best = word
        return best
